src/hotspot/share/gc/g1/g1UncommitRegionTask.cpp

```diff
@@ -129,7 +129,21 @@ void G1UncommitRegionTask::execute() {
   } else {
     // Nothing more to do, change state and report a summary.
     set_active(false);
+    _active_sem.signal();
     report_summary();
     clear_summary();
   }
 }
+
+void G1UncommitRegionTask::_wait_if_active() {
+  while (_active) {
+    _active_sem.wait();
+  }
+}
+
+void G1UncommitRegionTask::wait_if_active() {
```

Comment:
I guess this gets thread-safe because `_instance` is set in VM thread, and we're calling it from another VM operation. However a comment, and preferably an assertion on which thread is executing this would be helpful in the future.

Reply:
I have added there some assertions. Not sure they should stay this way, do you have more insight about it?

Reply:
Oh, I looked wrong, we're *not* calling it from VM operation (`VM_Crac::doit`). Asserting on the name of the thread makes no sense, that can be changed arbitrarily, and the checkpoint operation can be triggered from any thread, including a non-application thread if we use the `jcmd <pid> JDK.checkpoint`.

I haven't checked all code paths but it seems that the `_instance` is set from `VM_G1CollectFull`. If we are to skip synchronization of access to `_instance` we could call the `wait_if_active` from our `VM_Crac` and use `assert(Thread::current()->is_VM_thread())` as the assertion. However when we're in another VM operation I guess the GC thread we would be waiting for isn't suspended in a safepoint (so we would deadlock).

Reply:
It has been changed/simplified upon @AntonKozlov's comments so I find this discussion no longer relevant.